src/lib.rs

```diff
@@ -7,25 +7,30 @@ pub struct Pool<T> {
 }
 
 impl<T> Pool<T> {
-    pub fn new<C, D>(creation: C, clearance: D) -> Pool<T>
+    pub fn new<C, D>(cap: usize, creation: C, clearance: D) -> Pool<T>
     where
         C: Fn() -> T + 'static,
         D: Fn(&mut T) -> () + 'static,
     {
         Pool {
-            internal: Arc::new(Mutex::new(InternalPool::new(creation, clearance))),
+            internal: Arc::new(Mutex::new(InternalPool::new(cap, creation, clearance))),
         }
     }
 
     pub fn get<'a>(&'a self) -> ItemGuard<'a, T> {
         let mut pool = self.internal.lock().unwrap();
-        let item = if pool.free.is_empty() {
-            (*pool.creation)()
-        } else {
-            pool.free.pop().unwrap()
-        };
+        // If the pool is empty, we double the capacity and batch allocate
+        // empty elements.
+        if pool.free.is_empty() {
+            pool.capacity *= 2;
+            for _ in 0..(pool.capacity / 2) {
```

Comment:
Can't we also be smarter about how we grow `pool.free`?

Reply:
Did you have something particular in mind?

Generally doubling the size is considered a fine practice. Double the size when 100% saturated. Half the size when 25% saturated*. That way memory shouldn't fluctuate too much during runtime.

* We're currently not reclaiming memory, and that's probably okay (:

Reply:
We're doubling the capacity number but not actually increasing the size of the free buffer. That's done through the use of `push`. This _could_ lead to more than one allocation. Perhaps an optimization could be to increase the capacity of the underlying free pool buffer.

Reply:
yeah actually I think you're right, and we should track capacity of the vec directly on itself -- then upgrade that capacity and fill it. Think that would indeed speed things up.

Lemme try that real fast.

Reply:
okay done; numbers seem about the same -- a little bit lower, but it could just be noise.